integration/src/main/java/scripts/testscripts/DatasetPermissionOperations.java

```diff
@@ -0,0 +1,238 @@
+            .storageSourceId(snapshotId.toString())
+            .storageSystem(StorageSystem.TDR);
+    // Verify admin can create a dataset on the user snapshot
+    var datasetId = adminCreateDataset(request);
+    // But admin cannot access the underlying user snapshot
+    try {
+      adminSnapshotsApi.retrieveSnapshot(snapshotId, Collections.emptyList());
+      fail("Expected exception was not thrown");
+    } catch (Exception e) {
+      assertThat(
+          adminSnapshotsApi.getApiClient().getStatusCode(),
+          is(HttpStatusCodes.STATUS_CODE_UNAUTHORIZED));
+    }
+  }
+
+  private UUID adminCreateDataset(CreateDatasetRequest request) throws Exception {
+    var datasetId = adminDatasetsApi.createDataset(request).getId();
+    assertThat(adminDatasetsApi.getApiClient().getStatusCode(), is(HttpStatusCodes.STATUS_CODE_OK));
+    datasetIds.add(datasetId);
+    return datasetId;
+  }
+
+  private UUID createSnapshot(TestUserSpecification user) throws Exception {
+    var datarepoClient = new DatarepoClient(server, user);
+    var snapshotsApi = new SnapshotsApi(datarepoClient);
+    var request =
+        new SnapshotRequestModel()
+            .name("catalog_integration_test_" + System.currentTimeMillis())
+            .description("catalog test snapshot")
+            .profileId(defaultProfileId)
+            .addContentsItem(
+                new SnapshotRequestContentsModel()
+                    .datasetName(TEST_DATASET_NAME)
+                    .mode(SnapshotRequestContentsModel.ModeEnum.BYFULLVIEW));
+    var createJobId = snapshotsApi.createSnapshot(request).getId();
+
+    var jobApi = new JobsApi(datarepoClient);
+    JobModel job;
+    do {
+      job = jobApi.retrieveJob(createJobId);
+    } while (job.getJobStatus() == JobModel.JobStatusEnum.RUNNING);
+
+    if (job.getJobStatus() != JobModel.JobStatusEnum.SUCCEEDED) {
+      throw new RuntimeException("Create snapshot failed: " + job);
+    }
+    @SuppressWarnings("unchecked")
+    Map<Object, Object> result = (Map<Object, Object>) jobApi.retrieveJobResult(createJobId);
+    UUID snapshotId = UUID.fromString((String) result.get("id"));
+    log.info("created snapshot " + snapshotId);
+    snapshotIds.add(snapshotId);
+    return snapshotId;
+  }
+
+  @Override
+  public void cleanup(List<TestUserSpecification> testUsers) throws Exception {
+    if (!snapshotIds.isEmpty()) {
+      for (var snapshotId : snapshotIds) {
+        try {
+          adminSnapshotsApi.deleteSnapshot(snapshotId);
+        } catch (Exception e) {
+          var datarepoClient = new DatarepoClient(server, regularUser);
+          var snapshotsApi = new SnapshotsApi(datarepoClient);
+          snapshotsApi.deleteSnapshot(snapshotId);
+        } finally {
+          log.info("deleted snapshot " + snapshotId);
+        }
+      }
+    }
+    if (!datasetIds.isEmpty()) {
+      for (var datasetId : datasetIds) {
+        adminDatasetsApi.deleteDataset(datasetId);
```

Comment:
Wait I got this from SnapshotDatasetOperations:
```
  @Override
  public void cleanup(List<TestUserSpecification> testUsers) throws Exception {
    if (snapshotId != null) {
      snapshotsApi.deleteSnapshot(snapshotId);
      log.info("deleted snapshot " + snapshotId);
    }
    if (datasetId != null) {
      datasetsApi.deleteDataset(datasetId);
      log.info("deleted dataset " + datasetId);
    }
  }

```
which you wrote. 
Why is it different here?

Reply:
Yeah it's wrong, I thought I had fixed the order but I guess that change didn't get committed. It turns out either way will work because the auth check is done without calling TDR since the user is an admin user. But conceptually the dataset should be removed from catalog before the snapshot is deleted from TDR. If you could fix the other case too, that would be good.